File: tests/cip_configuration_reference.cpp
// Development-only direct calls to pinned native Configuration APIs.
#include "cip_graph_observation.h"
#include <GraphMol/CIPLabeler/configs/Tetrahedral.h>
#include <GraphMol/CIPLabeler/configs/Sp2Bond.h>
#include <GraphMol/CIPLabeler/configs/AtropisomerBond.h>
#include <GraphMol/CIPLabeler/rules/Rules.h>
#include <GraphMol/CIPLabeler/rules/Rule1a.h>
#include <GraphMol/CIPLabeler/rules/Rule1b.h>
#include <GraphMol/CIPLabeler/rules/Rule2.h>
#include <GraphMol/CIPLabeler/rules/Rule3.h>
#include <GraphMol/CIPLabeler/rules/Rule4a.h>
#include <GraphMol/CIPLabeler/rules/Rule4b.h>
#include <GraphMol/CIPLabeler/rules/Rule4c.h>
#include <GraphMol/CIPLabeler/rules/Rule5New.h>
#include <GraphMol/CIPLabeler/rules/Rule6.h>
#include <memory>

int main() {
  std::string line;
  while (std::getline(std::cin,line)) {
    try {
      std::istringstream request(line);
      int kind; request >> kind;
      if (kind==3) {
        unsigned a,b; request>>a>>b;
        std::vector<unsigned> first,second;
        for(unsigned i=0;i<4;++i) { first.push_back(a%4); a/=4; second.push_back(b%4); b/=4; }
        std::cout<<Configuration::parity4(first,second)<<std::endl;
        continue;
      }
      unsigned target,cfg,seed,full,aux,budget,repeat,reverse; int origin,nodeAtom,write;
      std::string hex;
      if(!(request>>target>>cfg>>origin>>nodeAtom>>seed>>full>>aux>>budget>>repeat>>reverse>>write>>hex))
        throw std::runtime_error("Invalid request");
      std::string binary;
      for(std::size_t i=0;i<hex.size();i+=2) binary.push_back(char(16*digit(hex[i])+digit(hex.at(i+1))));
      RDKit::RWMol source; RDKit::MolPickler::molFromPickle(binary,source); CIPMol mol(source);
      std::unique_ptr<Configuration> config;
      RDKit::Bond* bond=nullptr;
      RDKit::Atom* atom=nullptr;
      if(kind==0) { atom=mol.getAtom(target); config=std::make_unique<Tetrahedral>(mol,atom); }
      else {
        bond=mol.getBond(target); auto a=bond->getBeginAtom(); auto b=bond->getEndAtom();
        if(reverse) std::swap(a,b);
        if(kind==1) config=std::make_unique<Sp2Bond>(mol,bond,a,b,RDKit::Bond::BondStereo(cfg));
        else config=std::make_unique<AtropisomerBond>(mol,bond,a,b,RDKit::Bond::BondStereo(cfg));
      }
      std::unique_ptr<Digraph> external;
      if(origin>=0) external=std::make_unique<Digraph>(mol,mol.getAtom(origin),kind==2);
      auto &graph=external?*external:config->getDigraph();
      Registry registry; registry.add(graph.getOriginalRoot());
      Node* node=graph.getOriginalRoot();
      if(aux || nodeAtom>=0 || repeat) {
        graph.getNodes(mol.getAtom(mol.getNumAtoms()-1));
        View view(registry,graph.getOriginalRoot());
        if(aux) {
          for(std::size_t i=0;i<view.nodes.size();++i) {
            auto n=view.nodes[i];
            auto tag=n->getAtom()?n->getAtom()->getChiralTag():RDKit::Atom::CHI_UNSPECIFIED;
            auto desc=aux==3?(tag==RDKit::Atom::CHI_TETRAHEDRAL_CW?Descriptor::R:tag==RDKit::Atom::CHI_TETRAHEDRAL_CCW?Descriptor::S:Descriptor::NONE):aux==1 ? Descriptor((i+seed)%15) : Descriptor(3+(i+seed)%4);
            n->setAux(desc);
          }
          for(std::size_t i=0;i<view.edges.size();++i) view.edges[i]->setAux(aux==3?Descriptor::NONE:Descriptor((i*3+seed+1)%15));
          graph.setRule6Ref(mol.getAtom((target+1)%mol.getNumAtoms()));
        }
        if(nodeAtom>=0) {
          auto nodes=graph.getNodes(mol.getAtom(nodeAtom));
          std::vector<Node*> real;
          for(auto n:nodes) if(!n->isDuplicateOrH()) real.push_back(n);
          if(real.empty()) throw std::runtime_error("Missing auxiliary node");
          node=real.at(seed%real.size());
        }
        if(repeat) graph.changeRoot(view.nodes.at(seed%view.nodes.size()));
      }
      Rules rules({new Rule1a(),new Rule1b(),new Rule2()});
      if(full) { rules.add(new Rule3());rules.add(new Rule4a());rules.add(new Rule4b());rules.add(new Rule4c());rules.add(new Rule5New());rules.add(new Rule6()); }
      RDKit::RWMol empty; assignCIPLabels(empty,budget);
      std::ostringstream out; out<<"{\"foci\":[";
      for(std::size_t i=0;i<config->getFoci().size();++i) { if(i)out<<',';out<<config->getFoci()[i]->getIdx(); }
      out<<"],\"carriers\":[";
      for(std::size_t i=0;i<config->getCarriers().size();++i) { if(i)out<<',';auto a=config->getCarriers()[i];out<<(a?int(a->getIdx()):-1); }
      out<<"],\"results\":[";
      for(unsigned pass=0;pass<(repeat?2u:1u);++pass) {
        if(pass)out<<',';
        auto desc=origin>=0 || nodeAtom>=0 ? config->label(node,graph,rules) : config->label(rules);
        auto chosen=write>=0?Descriptor(write):(kind==0?Descriptor::R:kind==1?Descriptor::E:Descriptor::M);
        auto hasBefore=config->hasPrimaryLabel();
        config->setPrimaryLabel(chosen);
        std::vector<unsigned> ranks; std::string code;
        const RDKit::RDProps& props=kind==0?static_cast<const RDKit::RDProps&>(*atom):static_cast<const RDKit::RDProps&>(*bond);
        props.getProp(RDKit::common_properties::_CIPNeighborOrder,ranks);
        props.getProp(RDKit::common_properties::_CIPCode,code);
        auto hasSet=config->hasPrimaryLabel(); config->resetPrimaryLabel();
        out<<"{\"descriptor\":"<<int(desc)<<",\"code\":\""<<code<<"\",\"neighbor_order\":[";
        for(std::size_t i=0;i<ranks.size();++i) { if(i)out<<',';out<<(ranks[i]==IMPLICITH?-1:static_cast<long long>(ranks[i])); }
        out<<"],\"has\":["<<(hasBefore?"true":"false")<<','<<(hasSet?"true":"false")<<','<<(config->hasPrimaryLabel()?"true":"false")<<"],\"stereo\":";
        if(bond) {
          out<<'['<<int(bond->getStereo())<<",[";
          for(std::size_t i=0;i<bond->getStereoAtoms().size();++i) { if(i)out<<',';out<<bond->getStereoAtoms()[i]; }
          out<<"]]";
        } else out<<"null";
        out<<",\"graph\":"; View view(registry,graph.getOriginalRoot()); view.write(out,registry,graph,mol); out<<'}';
      }
      out<<"]}";std::cout<<out.str()<<std::endl;
    } catch(const MaxIterationsExceeded&) { std::cout<<"{\"error\":\"iterations\"}"<<std::endl;
    } catch(const TooManyNodesException&) { std::cout<<"{\"error\":\"nodes\"}"<<std::endl;
    } catch(const std::exception&) { std::cout<<"{\"error\":\"invalid\"}"<<std::endl; }
  }
}

File: tests/cip_configuration_reference.py
"""Direct pinned-native configuration labels, carriers, mutations and graphs."""

import argparse
import contextlib
import gzip
import hashlib
import itertools
import json
import random
import subprocess
from pathlib import Path
from typing import TYPE_CHECKING

from rdkit import Chem, RDLogger, rdBase

if TYPE_CHECKING or __package__:
    from .cip_molecule_reference import molecules
    from .perception_reference import snapshot
else:
    from cip_molecule_reference import molecules
    from perception_reference import snapshot

FIXTURE = Path(__file__).parent / "fixtures/cip-configuration-native.json.gz"
KEYS = (
    "kind",
    "target",
    "cfg",
    "origin",
    "node_atom",
    "seed",
    "full",
    "aux",
    "budget",
    "repeat",
    "reverse",
    "write",
)


def inputs():
    texts = [
        "N[C@@H](C)C(=O)O",
        "F[C@](Cl)(Br)I",
        "[1H][C@H](F)Cl",
        "C[S@H]=O",
        "C[S@](=O)CC",
        "C[P@](F)Cl",
        "C[C@H](C)O",
        "C[C@](C)(C)C",
        "C[C@](CC)(CC)C",
        "[C@]12(CCCCC1)CCCCC2",
        "C[C@H]1CCCCO1",
        "F/C=C/F",
        "Cl/C(F)=C(/Br)I",
        "[1H]/C=C/F",
        "C/C=C/C=C/C",
        "C/C=C\\C",
        "F/C=C(/[C@H](F)Cl)[C@@H](F)Cl",
        "CC(Cl)=C(C)CC",
        "CC(F)C(C)C(Br)C",
        "c1ccccc1-c1ncccc1",
        "Cc1cccc(Br)c1-c1c(Cl)cccc1F",
        "CC",
        "C1C2CC3CC1CC(C2)C3",
        "[C@H](F)Cl",
        "[C@](F)(Cl)Br",
        "[P@H](F)Cl",
        "[C@H3]F",
        "[C@]",
        "[C@](F)(F)(F)F",
        "FC(F)C(Cl)(Br)C(F)F",
        "F[C@]([C@H](F)Cl)([C@@H](F)Cl)Br",
        "C[C@]([C@H](F)Cl)([C@@H](F)Cl)C",
        "F[C@H](Cl)C(C)=C([C@H](F)Cl)[C@@H](F)Cl",
        "F[C@H](Cl)C([C@@H](F)Cl)=C([C@H](F)Cl)[C@@H](F)Cl",
        "F[C@H](Cl)C(C)C([C@H](F)Cl)[C@@H](F)Cl",
        "F[C@H](Cl)C([C@@H](F)Cl)C([C@H](F)Cl)[C@@H](F)Cl",
    ]
    rng = random.Random(51045)
    for i, text in enumerate(texts):
        params = Chem.SmilesParserParams()
        params.removeHs = False
        params.sanitize = False
        mol = Chem.MolFromSmiles(text, params)
        assert mol is not None
        try:
            Chem.SanitizeMol(mol)
        except (ValueError, RuntimeError):
            mol.UpdatePropertyCache(strict=False)
        Chem.GetSymmSSSR(mol)
        # Preserve low-level tags that a complete stereo cleanup may remove.
        yield f"special/{i}", mol, "symmetric"
        order = list(range(mol.GetNumAtoms()))
        rng.shuffle(order)
        yield f"permuted/{i}", Chem.RenumberAtoms(mol, order), "symmetric"
    for name, mol, cache in molecules():
        if name.startswith("molecule/") and int(name.split("/")[1]) < 160:
            yield name, mol, cache


def configurations():
    for name, source, cache in inputs():
        targets = [(0, a.GetIdx()) for a in source.GetAtoms() if a.GetDegree() in (2, 3, 4)]
        targets += [
            (1, b.GetIdx()) for b in source.GetBonds() if b.GetBondType() == Chem.BondType.DOUBLE
        ]
        targets += [
            (2, b.GetIdx())
            for b in source.GetBonds()
            if b.GetBondType() == Chem.BondType.SINGLE
            and b.GetBeginAtom().GetDegree() > 1
            and b.GetEndAtom().GetDegree() > 1
        ]
        if name.startswith("molecule/"):
            targets = targets[:3]
        for kind, target in targets:
            mol = Chem.Mol(source)
            cfg = 1
            if kind == 0:
                atom = mol.GetAtomWithIdx(target)
                if atom.GetChiralTag() not in (
                    Chem.ChiralType.CHI_TETRAHEDRAL_CW,
                    Chem.ChiralType.CHI_TETRAHEDRAL_CCW,
                ):
                    atom.SetChiralTag(Chem.ChiralType.CHI_TETRAHEDRAL_CW)
                focus = target
            else:
                bond = mol.GetBondWithIdx(target)
                focus = bond.GetBeginAtomIdx()
                ends = [bond.GetBeginAtom(), bond.GetEndAtom()]
                nbrs = [
                    [a.GetIdx() for a in end.GetNeighbors() if a.GetIdx() != ends[1 - i].GetIdx()]
                    for i, end in enumerate(ends)
                ]
                if kind == 1:
                    if not all(nbrs):
                        continue
                    bond.SetStereoAtoms(nbrs[0][0], nbrs[1][0])
                    cfg = 4 + target % 2
                    bond.SetStereo(Chem.BondStereo.values[cfg])
                else:
                    cfg = 6 + target % 2
                    bond.SetStereo(Chem.BondStereo.values[cfg])
            yield name, mol, cache, kind, target, cfg, focus


def cases():
    for index, (name, mol, cache, kind, target, cfg, focus) in enumerate(configurations()):
        variants = range(9) if name.startswith(("special/", "permuted/")) else range(3)
        for variant in variants:
            changed = Chem.Mol(mol)
            if variant == 4:
                obj = (
                    changed.GetAtomWithIdx(target) if kind == 0 else changed.GetBondWithIdx(target)
                )
                obj.SetProp("_CIPCode", "legacy")
            reverse = variant in (6, 8) and kind != 0
            origin = (focus + 1) % changed.GetNumAtoms() if variant in (2, 3) else -1
            node_atom = focus if origin >= 0 else -1
            if variant == 3 and kind != 0:
                node_atom = changed.GetBondWithIdx(target).GetEndAtomIdx()
            yield (
                changed,
                dict(
                    name=f"{name}/{kind}/{target}/{variant}",
                    before=snapshot(changed, cache),
                    kind=kind,
                    target=target,
                    cfg=cfg,
                    origin=origin,
                    node_atom=node_atom,
                    seed=index * 31 + variant * 7,
                    full=variant != 0,
                    aux=3 if variant >= 7 else 1 if variant == 3 else 2 if variant in (4, 6) else 0,
                    budget=1 if variant == 5 else 3000,
                    repeat=variant in (1, 4),
                    reverse=reverse,
                    write=-1,
                ),
            )
    # Explicit constructor and primary-label rejection, missing inferred controls,
    # and native implicit-H/phantom centers with too few incident graph edges.
    for kind, text, target in (
        (0, "[C]", 0),
        (0, "CF", 0),
        (0, "C(F)Cl", 0),
        (1, "FC=CF", 1),
        (2, "CC", 0),
    ):
        for variant in range(20):
            mol = Chem.MolFromSmiles(text)
            assert mol is not None
            cfg = 4 if kind == 1 else 6
            if kind == 0:
                mol.GetAtomWithIdx(target).SetChiralTag(Chem.ChiralType.CHI_TETRAHEDRAL_CW)
            elif kind == 1:
                bond = mol.GetBondWithIdx(target)
                bond.SetStereo(Chem.BondStereo.STEREOE)
                if variant % 3:
                    for atom in mol.GetAtoms():
                        atom.SetUnsignedProp("_CIPRank", atom.GetIdx())
            yield (
                mol,
                dict(
                    name=f"invalid/{kind}/{text}/{variant}",
                    before=snapshot(mol, "symmetric"),
                    kind=kind,
                    target=target,
                    cfg=cfg,
                    origin=-1,
                    node_atom=-1,
                    seed=0,
                    full=True,
                    aux=0,
                    budget=1000,
                    repeat=False,
                    reverse=False,
                    write=variant % 18,
                ),
            )


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--oracle", type=Path)
    parser.add_argument("--write-fixture", action="store_true")
    args = parser.parse_args()
    RDLogger.DisableLog("rdApp.*")
    assert rdBase.rdkitVersion == "2026.03.6"
    if args.write_fixture and not args.oracle:
        parser.error("--write-fixture requires --oracle")
    records = {} if args.write_fixture else json.loads(gzip.decompress(FIXTURE.read_bytes()))
    with contextlib.ExitStack() as stack:
        child = None
        if args.oracle:
            child = stack.enter_context(
                subprocess.Popen(
                    [str(args.oracle)],
                    stdin=subprocess.PIPE,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.DEVNULL,
                    text=True,
                )
            )
            stack.callback(child.kill)
        if not args.write_fixture:
            print(json.dumps(dict(rdkit_version=rdBase.rdkitVersion)))
        for mol, case in cases():
            name = case["name"]
            digest = hashlib.sha256(
                json.dumps(case, sort_keys=True, separators=(",", ":")).encode()
            ).hexdigest()
            if child:
                assert child.stdin and child.stdout
                params = " ".join(str(int(case[key])) for key in KEYS)
                child.stdin.write(
                    f"{params} {mol.ToBinary(Chem.PropertyPickleOptions.AllProps).hex()}\n"
                )
                child.stdin.flush()
                line = child.stdout.readline()
                if not line:
                    assert child.wait() < 0, name
                    expected = {"error": "native_crash"}
                    child = stack.enter_context(
                        subprocess.Popen(
                            [str(args.oracle)],
                            stdin=subprocess.PIPE,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.DEVNULL,
                            text=True,
                        )
                    )
                    stack.callback(child.kill)
                else:
                    expected = json.loads(line)
                if args.write_fixture:
                    records[name] = [digest, expected]
                    continue
            else:
                recorded, expected = records.pop(name)
                assert recorded == digest, f"Native input changed: {name}"
            print(json.dumps({**case, "expected": expected}))
        # Exhaustively include repeated elements as well as all permutations.
        parity = []
        for a, b in itertools.product(range(256), repeat=2):
            if child:
                assert child.stdin and child.stdout
                child.stdin.write(f"3 {a} {b}\n")
                child.stdin.flush()
                parity.append(int(child.stdout.readline()))
        if args.write_fixture:
            records["parity"] = parity
        else:
            print(json.dumps(dict(parity=parity if child else records.pop("parity"))))
        if child:
            assert child.stdin
            child.stdin.close()
            assert child.wait() == 0
    if args.write_fixture:
        FIXTURE.write_bytes(
            gzip.compress(json.dumps(records, separators=(",", ":")).encode(), mtime=0)
        )
        print(
            f"Recorded {len(records) - 1} native configurations and {len(records['parity'])} parities ({FIXTURE.stat().st_size} bytes)"
        )
    elif not args.oracle:
        assert not records, "Unused native fixtures"


if __name__ == "__main__":
    main()
